[MRG+1] Adding a 4Dplot function displaying as a contours in img_plotting

KamalakerDadi opened the pull request:
Edited by @AlexandreAbraham
- [x] thicken contour lines
- [x] add an example of plotting with several atlases
- [x] make code PEP8 compliant
- [x] for thresholding, accept one value, or a list of values, or a percentage, or a list of percentage

I have basically utilised the content of the code and documentation of each arguments for writing a function called ```plot_4d``` in ```img_plotting```.
Here is the code snippet to call the function ```plot_4d``` for plotting in contours. 
Harvard Oxford Atlas:
```python
from nilearn import datasets, plotting
import matplotlib.pyplot as plt
atlas_img_4D, labels = datasets.fetch_harvard_oxford('cort-prob-2mm')
display = plotting.img_plotting.plot_4d(atlas_img_4D)
plt.show()
```
Result:
![figure_1](https://cloud.githubusercontent.com/assets/11410385/7667394/2952f1ae-fc06-11e4-8055-70f6ba96a2b2.png)


MSDL Atlas:
```python
from nilearn import datasets, plotting
import matplotlib.pyplot as plt
atlas_data = datasets.fetch_msdl_atlas()
atlas_filename = atlas_data.maps
display = plotting.img_plotting.plot_4d(atlas_filename)
plt.show()
```
Result:
![figuremsdl](https://cloud.githubusercontent.com/assets/11410385/7667400/54c4ae68-fc06-11e4-8a05-cfdb56c63ad8.png)

KamalakerDadi commented:
issue #588

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -313,6 +314,78 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
 # Usage-specific functions
 
 
+def plot_4d(img, anat_img=MNI152TEMPLATE, cut_coords=None,
```
The name of the function is a bit misleading. Given that EPI is 4d, I may want to put it in this function, which is not the intended purpose. Something like `plot_atlas` or `plot_contours` would be more meaningful I think.

KamalakerDadi replied:
Yes, I agree. I will change it.

GaelVaroquaux replied:
> > +def plot_4d(img, anat_img=MNI152TEMPLATE, cut_coords=None,

> Yes, I agree. I will change it.

It could be called 'plot_multi_maps', and have a contours option to
switch between contours and continuous colors.

AlexandreAbraham replied:
Why not `plot_maps`?

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -313,6 +314,78 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    vmax = kwargs.pop('vmax', vmax)
+    img = _utils.check_niimg_4d(img)
+    n_colors = img.shape[3]
+    node_colors = [plt.cm.spectral(i / float(n_colors))
```
The cmap of the contours should be an option.

KamalakerDadi replied:
To be clarified myself, should be added as an optional argument ?

KamalakerDadi replied:
One thing I saw is that if I use default ```cmap``` which was already defined. I don't see any contours in colors instead showed in a gray scale. I thought there would be double mismatch with cmap.

GaelVaroquaux replied:
> > +    node_colors = [plt.cm.spectral(i / float(n_colors))

> To be clarified myself, should be added as an optional argument ?

Yes, the cmap should be there as an argument.

AlexandreAbraham replied:
You already have `cmap` parameter and you use it for the anatomical background. You should rely on default argument for the anat and use the cmap arg here. In my experience, gist_rainbow is better than spectral since spectral have black and white colors.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -313,6 +314,78 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+                       draw_cross=draw_cross, black_bg=black_bg,
+                       vmin=vmin, vmax=vmax, cmap=cmap, **kwargs)
+    for i, cur_img in enumerate(iter_img(img)):
+        display.add_contours(cur_img, levels=[30], linewidths=2, 
```
The value of levels is arbitrary here.

AlexandreAbraham replied:
Note: this is why you have nothing when plotting MSDL. We should put it as a parameter and have by default an 'auto' strategy where we use `compute_background_mask.`

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -313,6 +314,78 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    """
+    anat_img, black_bg, vmin, vmax = _load_anat(anat_img,
+                                                dim=dim, black_bg=black_bg)
+    vmin = kwargs.pop('vmin', vmin)
```
Is there a sense in applying vmin and vmax on anat? I would rather have `anat_vmin` and `anat_vmax` and maybe a `threshold` parameter to be considered as level.

KamalakerDadi replied:
I have looked how ```plot_anat``` and included here. Do you mean it will be nice to use keywords ```anat_vmin``` instead of ```vmin``` ?
I can also check ```threshold`` parameter.

AlexandreAbraham replied:
> I have looked how plot_anat and included here. Do you mean it will be nice to use keywords anat_vmin instead of vmin ?

Well, take example on `plot_stat_map`. There is no option to set vmin and vmax for the anat. I think that you should do the same.

> I can also check `threshold parameter.

Yes, I think that it would be useful. Or maybe vmin to stick the conventions. @GaelVaroquaux, any thoughts on this?

KamalakerDadi replied:
At the moment, I used ```display.add_contours(cur_img, levels=[anat_vmin, anat_vmax])```

AlexandreAbraham replied:
Why do use `anat_vmin` to display non anatomic data? Plus, why would you put two contours in the figure?

AlexandreAbraham commented:
I am wondering if we want to have such a function instead of educating the user about how to display several regions himself. The only additional logic in this function is the way the colormap is computed, which is a pain in matplotlib but for which we can provide a utility function.

I have one myself and so to do the same thing I write:

```python
p = plot_anat(...)

for region_img, color in zip(iter_img(atlas_img), get_colors('gist_rainbow', len(atlas_img.shape[-1]))):
    p.add_countours(region_img, ...)
```

What would be (in my sense) a good addition for such function:
* a "group" heuristic for `find_cut_coords` because it will always be centered here I think (but this may be something to add in the slicer itself, an update of the cut coords when contours or overlays are added)
* an intelligent coloring code. Here, neighboring regions have similar colors because the atlas is made this way. We should write an algorithm to have maximum contrast between neigboring regions, this is a pain but this would be the new killer feature for me
* same code but with `add_overlay` to display several stat maps.

GaelVaroquaux commented:
> I am wondering if we want to have such a function instead of educating
> the user about how to display several regions himself.

This will provide a lot of value for beginners. It's a feature that many
people have been curious about. Also, many of us have a copy-pasted
variant of this code. Getting is right is somewhat challenging, and each
time someone rewrites it, its time lost.

When something keeps coming up in people's code, it's often a good thing
to make a function out of it. Also, we are going to need it to make the
output of the ICA.

AlexandreAbraham commented:
Well, in that case, I would say that we need to support maps and labels. Maybe an `iter_map` function would be useful. It would iterate on 4d if there is one, and if it is 3D with integer values (ie labels), it would return a binary mask corresponding to each label.

GaelVaroquaux commented:
I would provide 2 functions, and not do magic, but have a good error
message. We already have the first function: plot_roi.

KamalakerDadi commented:
I am thinking more or less like a enhancement to plotting functions, for example 4D. As far as I have seen, there is no specific example describing plotting 4D. As a users point, I thought it would be nice to input data and view the maps as an overlay or contours on the anatomical image.

KamalakerDadi commented:
Examples with Harvard Oxford, MSDL atlases: default ```threshold``` mode
```python
from nilearn import datasets, plotting
# Harvard Oxford Atlas
atlas_img_4D = datasets.fetch_harvard_oxford('cort-prob-2mm')
# Visualization of Harvard Oxford Atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], 
                                                title='Harvard-Oxford with contours')
# Visualization of Harvard Oxford Atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], view_type='continuous',
                                                title='Harvard-Oxford with continuous')
```
Output:
![hocontours](https://cloud.githubusercontent.com/assets/11410385/7909869/ee77dd96-084f-11e5-8ac0-145d487b6aa7.png)

![hocontinuous](https://cloud.githubusercontent.com/assets/11410385/7909871/f46734d6-084f-11e5-8ea6-37cfade51177.png)

MSDL Atlas Example:
```python
from nilearn import datasets, plotting
# MSDL Atlas
atlas_data = datasets.fetch_msdl_atlas()
atlas_filename = atlas_data.maps
# Visualization of MSDL Atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(atlas_filename,
                                                title='MSDL with contours')
# Visualization of MSDL Atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(atlas_filename, view_type='continuous',
                                                title='MSDL with continuous')
```
Output:
![msdlcontours](https://cloud.githubusercontent.com/assets/11410385/7909892/28d93494-0850-11e5-9071-517163b27955.png)
![msdlcontinuous](https://cloud.githubusercontent.com/assets/11410385/7909893/2d704934-0850-11e5-9224-abae23e8a240.png)

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -30,12 +31,32 @@
 ################################################################################
 # Core, usage-agnostic functions
 
+def black_to(color):
```
Small remark, I think that @GaelVaroquaux wanted colormaps going from transparent to color and not from black to color.

GaelVaroquaux replied:
Indeed.

GaelVaroquaux replied:
> Small remark, I think that @GaelVaroquaux wanted colormaps going from
> transparent to color and not from black to color.

Yes. And we have a function for that in plotting.com, I believe.

AlexandreAbraham replied:
I has been fixed.

On Fri, Jun 12, 2015 at 9:33 PM, Gael Varoquaux <notifications@github.com>
wrote:

> In nilearn/plotting/img_plotting.py
> <https://github.com/nilearn/nilearn/pull/589#discussion_r32347171>:
>
> >  from .displays import get_slicer, get_projector
> >  from . import cm
> >
> >  ################################################################################
> >  # Core, usage-agnostic functions
> >
> > +def black_to(color):
>
>  Small remark, I think that @GaelVaroquaux
> <https://github.com/GaelVaroquaux> wanted colormaps going from
> transparent to color and not from black to color.
> Yes. And we have a function for that in plotting.com, I believe.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/589/files#r32347171>.
>

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +333,116 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+        cont = ListedColormap(color).colors
+        if threshold is not None:
+           if hasattr(threshold, '__iter__'):
+              value = threshold[i]
```
The best way to handle that is to put `threshold` in the zip. And do this test before:
```python
if not hasattr(threshold, '__iter__'):
    threshold = [threshold] * n_maps

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +333,116 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+                value = threshold
+        # To threshold or choose the level of the contours
+        percentile = 100 * (1 - (value/n_maps))
+        data = np.abs(map_img.get_data())
```
You don't need to take the abs. It's already done in `fast_abs_percentile` (hence the `abs` in the name ;)).

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +333,116 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+        data = np.abs(map_img.get_data())
+        level = fast_abs_percentile(data, percentile) + 1e-4
+        
+        if view_type == 'auto': 
```
Where is the `view_type` contours?

KamalakerDadi replied:
```view_type``` contours does it in default ```auto``` mode.
You mean to switch to display only as contours without any contour fillings ?

AlexandreAbraham replied:
What we said, but maybe @GaelVaroquaux wan confirm, is that we have three view types:
* contours which plot contours only
* continuous which plot the statistical maps
* auto that chooses between these 2 modes depending on some heuristic on the data (typically, we can plot continuous maps if there is not too much overlap between the maps)

GaelVaroquaux replied:
Yes. And contours should be filled contours, like what Yannick does. Filled contours are not continous overlays, they are obtained using the matplotlib contourf function.

GaelVaroquaux replied:
> What we said, but maybe @GaelVaroquaux wan confirm, is that we have three view
> types:

>   • contours which plot contours only
>   • continuous which plot the statistical maps
>   • auto that chooses between these 2 modes depending on some heuristic on the
>     data (typically, we can plot continuous maps if there is not too much
>     overlap between the maps)

Yes. I'd also like a third option 'filled_contours', that reproduces the
figures from Yannick's work. This one would not be turned on
automatically.

AlexandreAbraham replied:
For the moment, you can do it by selecting view_type='contour' and
filled=True.

On Fri, Jun 12, 2015 at 9:34 PM, Gael Varoquaux <notifications@github.com>
wrote:

> In nilearn/plotting/img_plotting.py
> <https://github.com/nilearn/nilearn/pull/589#discussion_r32347297>:
>
> > +    color_list = color_map(np.linspace(0, 1, n_maps))
> > +
> > +    for i, (map_img, color) in enumerate(zip(iter_img(maps_img), color_list)):
> > +        over = black_to(color)
> > +        cont = ListedColormap(color).colors
> > +        if threshold is not None:
> > +           if hasattr(threshold, '__iter__'):
> > +              value = threshold[i]
> > +           else:
> > +                value = threshold
> > +        # To threshold or choose the level of the contours
> > +        percentile = 100 * (1 - (value/n_maps))
> > +        data = np.abs(map_img.get_data())
> > +        level = fast_abs_percentile(data, percentile) + 1e-4
> > +
> > +        if view_type == 'auto':
>
>  What we said, but maybe @GaelVaroquaux <https://github.com/GaelVaroquaux>
> wan confirm, is that we have three view types: • contours which plot
> contours only • continuous which plot the statistical maps • auto that
> chooses between these 2 modes depending on some heuristic on the data
> (typically, we can plot continuous maps if there is not too much overlap
> between the maps)
> Yes. I'd also like a third option 'filled_contours', that reproduces the
> figures from Yannick's work. This one would not be turned on automatically.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/589/files#r32347297>.
>

GaelVaroquaux replied:
Should we remove the "filled" argument, and add an additional option to the view_type? I think that avoiding the multiplicity of kwargs is a good thing. 

Sent from my phone. Please forgive brevity and mis spelling



On Jun 12, 2015, 14:01, at 14:01, Alexandre Abraham <notifications@github.com> wrote:
>> +    color_list = color_map(np.linspace(0, 1, n_maps))
>> +
>> +    for i, (map_img, color) in enumerate(zip(iter_img(maps_img),
>color_list)):
>> +        over = black_to(color)
>> +        cont = ListedColormap(color).colors
>> +        if threshold is not None:
>> +           if hasattr(threshold, '__iter__'):
>> +              value = threshold[i]
>> +           else: 
>> +                value = threshold
>> +        # To threshold or choose the level of the contours
>> +        percentile = 100 * (1 - (value/n_maps))
>> +        data = np.abs(map_img.get_data())
>> +        level = fast_abs_percentile(data, percentile) + 1e-4
>> +        
>> +        if view_type == 'auto': 
>
>For the moment, you can do it by selecting view_type='contour' and
>filled=True.
>
>On Fri, Jun 12, 2015 at 9:34 PM, Gael Varoquaux
><notifications@github.com>
>wrote:
>
>> In nilearn/plotting/img_plotting.py
>> <https://github.com/nilearn/nilearn/pull/589#discussion_r32347297>:
>>
>> > +    color_list = color_map(np.linspace(0, 1, n_maps))
>> > +
>> > +    for i, (map_img, color) in enumerate(zip(iter_img(maps_img),
>color_list)):
>> > +        over = black_to(color)
>> > +        cont = ListedColormap(color).colors
>> > +        if threshold is not None:
>> > +           if hasattr(threshold, '__iter__'):
>> > +              value = threshold[i]
>> > +           else:
>> > +                value = threshold
>> > +        # To threshold or choose the level of the contours
>> > +        percentile = 100 * (1 - (value/n_maps))
>> > +        data = np.abs(map_img.get_data())
>> > +        level = fast_abs_percentile(data, percentile) + 1e-4
>> > +
>> > +        if view_type == 'auto':
>>
>>  What we said, but maybe @GaelVaroquaux
><https://github.com/GaelVaroquaux>
>> wan confirm, is that we have three view types: • contours which plot
>> contours only • continuous which plot the statistical maps • auto
>that
>> chooses between these 2 modes depending on some heuristic on the data
>> (typically, we can plot continuous maps if there is not too much
>overlap
>> between the maps)
>> Yes. I'd also like a third option 'filled_contours', that reproduces
>the
>> figures from Yannick's work. This one would not be turned on
>automatically.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/nilearn/nilearn/pull/589/files#r32347297>.
>>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nilearn/nilearn/pull/589/files#r32354591

AlexandreAbraham commented:
> As I see I missed the "label = 0", which is "Frontal Pole"

I don't see how you can lose a region with this threshold strategy. Could you add some debug info, like the min and max of the frontal pole map and the computed threshold?

AlexandreAbraham commented:
> To avoid maximum overlaps between each maps, I used less value ```threshold=1e-1```.

For me, this value has no sense. What we said about threshold:
* if threshold is a string, it must look like "38%" and, in that case, we keep all the values that are, in absolute, above the threshold of 38%
* if threshold if a float, we use this value as a threshold
* if threshold is a list, then the same rules as above apply but each map has a different threshold

You can take a look at the parameter `edge_threshold` of `plot_connectome` which is handled the same way.

KamalakerDadi commented:
After addressing some comments: 
**Fixing contour fillings using contourf**
Output: 
```python
import matplotlib.pyplot as plt
from nilearn import datasets, plotting

# Harvard Oxford Atlas
atlas_img_4D = datasets.fetch_harvard_oxford('cort-prob-2mm')
# Visualization of Harvard Oxford Atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0],
                                                title='Harvard-Oxford with contours')
# Visualization of Harvard Oxford Atlas maps as a contour fillings type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], fill=True,
                                                title='Harvard-Oxford with contour fillings')
# Visualization of Harvard Oxford Atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], view_type='continuous',
                                                title='Harvard-Oxford with continuous')
plt.show()
```
**contours**
**continuous overlays**
 **contours with fillings**
![hocontours](https://cloud.githubusercontent.com/assets/11410385/8058271/9ab3b904-0eb8-11e5-8954-1ec02903b50b.png)
![hooverlays](https://cloud.githubusercontent.com/assets/11410385/8058277/a173e3c2-0eb8-11e5-942b-802e607aa45a.png)
![hocontourfillings](https://cloud.githubusercontent.com/assets/11410385/8058278/a492db3a-0eb8-11e5-8f6a-a6677fd4abe0.png)

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -88,6 +88,9 @@ def draw_2d(self, data_2d, data_bounds, bounding_box,
         ax = self.ax
+        # Switches into contour fillings by overriding type as 'contourf'
+        if 'fill' in kwargs and kwargs['fill'] is True:
+            type = 'contourf'
```
You should not add a parameter but directly use `type='contourf'` in function call.

KamalakerDadi replied:
If I add ```type='contourf'```, will it not disturb other contour plotting examples ?
Did I understood your comment wrongly?

AlexandreAbraham replied:
You should add that in the function call. Instead of `draw_2d(..., fill=True)`, you can write `draw_2d(..., type="contourf")`.

AlexandreAbraham replied:
You did not address this comment.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -95,7 +98,7 @@ def draw_2d(self, data_2d, data_bounds, bounding_box,
 
         self.add_object_bounds((xmin_, xmax_, zmin_, zmax_))
-
+      
```
You should not add spaces in these lines.

KamalakerDadi replied:
Not done intentionally. It happened when I was looking into the code whats happening to know how the add_contour does.

AlexandreAbraham replied:
I know it's not intentional ;). You should take a look at vim-flake8 that spot these problems.

AlexandreAbraham replied:
You did not address this comment.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -562,14 +565,14 @@ def _map_show(self, img, type='imshow', resampling_interpolation='continuous', *
+        
         ims = []
-        to_iterate_over = zip(self.axes.values(), data_2d_list)
+        to_iterate_over = zip(self.axes.values(), data_2d_list)        
```
Trailing whitespaces.

KamalakerDadi replied:
Same not done intentionally.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -30,12 +31,26 @@
+        'alpha', cmapspec, plt.cm.LUTSIZE)
+    cmap._init()
+    cmap._lut[:, -1] = np.linspace(.7, .95, cmap._lut.shape[0])
+    cmap._lut[-1, -1] = 0
```
Could you add a comment about what happens here because I don't get it :confused:

KamalakerDadi replied:
Even me too. I am sorry but basically for an easy way I took code snippet from Yannick. May be I will understand now to answer.

AlexandreAbraham replied:
You did not address this comment. What you can do is directly ask yannick or even do a `git blame` where you found the code and look at the log of the corresponding commit. It may be helpful.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,135 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    if threshold is None:
+        # percentage is predefined at the moment, 
+        # for a nicer look avoiding maximum overlaps
+        threshold = [99.7] * n_maps
```
You still don't handle the case where threshold is a precise value and not a percentile.

KamalakerDadi replied:
At the moment, I used simply a value. I would add code validations to accept string (percentage) as done in ```plot_connectome``` once we have a good discussions and conclusions on using a  ```threshold```

AlexandreAbraham replied:
We already had this discussion with Gael. You can code it, I think that's all what's left in this PR.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,135 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+            display.add_contours(map_img, levels=[level],
+                                 linewidths=linewidths,
+                                 colors=[color])
+            if fill:
```
I think `filled` would be better. Maybe we should add a `'filled_contours'` `view_type`, what do you think?

KamalakerDadi replied:
```if filled``` would be better but regarding ```'filled_contours'``` are we not trying to confuse by increasing the options ?

AlexandreAbraham replied:
Obviously we should only chose one of these solutions. I asked Gael on the PR discussion, we'll wait for his answer.

AlexandreAbraham commented:
@GaelVaroquaux I'm not convinced by the filled contours and I think that we should let the user decide if he wants them or not. Do you agree with that? If so, should we add a `view_type` or a kwarg?

GaelVaroquaux commented:
I agree that we should let the option open. It can be useful, as can be seen on figures from Yannick's paper.

Sent from my phone. Please forgive brevity and mis spelling



On Jun 9, 2015, 07:02, at 07:02, Alexandre Abraham <notifications@github.com> wrote:
>@GaelVaroquaux I'm not convinced by the filled contours and I think
>that we should let the user decide if he wants them or not. Do you
>agree with that? If so, should we add a `view_type` or a kwarg?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nilearn/nilearn/pull/589#issuecomment-110365075

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    if hasattr(threshold, '__iter__'):
+        if len(threshold) != n_maps:
+            raise TypeError('The list of values to threshold '
+                                'should be equal to number of maps')
```
Put a message that is more precise here.

Something like:
```python
raise TypeError('You tried to plot an atlas with %i maps but gave %i values for thresholding, which is inconsistent.' % (n_maps, len(threshold)))
```

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    if threshold is None:
+        # it will use default percentage, 
+        # for a nicer look avoiding maximum overlaps for visualization
+        threshold = ["99.7%"] * n_maps
```
```
threshold = "99.7%"
```

Would work too and is clearer in this context.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+            except ValueError as exc:
+                exc.args += (message, )
+                raise
+            thr = fast_abs_percentile(data, percentile) + 1e-4
```
When you add magic values, specify why you do that. Plus, this value is not small.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+                raise
+            thr = fast_abs_percentile(data, percentile) + 1e-4
+        elif False: # TODO: check if it's not a number
+            raise TypeError('Threshold must be a number or...')
```
TODO here.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+                # Append the lower boundary value as 0 for contour fillings
+                display.add_contours(map_img, levels=[level, 0.],
+                                     linewidths=linewidths,
+                                     colors=[color[:3].tolist() + [.5]],
```
Add an inline comment like `# .5 alpha for filling color`

AlexandreAbraham commented:
You still have cosmetics details to solve. Apart from that, the code is ready.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -30,12 +31,26 @@
 ################################################################################
 # Core, usage-agnostic functions
 
+def _alpha_cmap(color):
```
This function is already in cm.py. Please reuse it.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -312,6 +327,155 @@ def _load_anat(anat_img=MNI152TEMPLATE, dim=False, black_bg='auto'):
+    anat_img, black_bg, _, _ = _load_anat(anat_img, dim=dim,
+                                          black_bg=black_bg)
+  
+    display = plot_img(anat_img, cut_coords=cut_coords,
```
Should we be using plot_anat here?

KamalakerDadi commented:
@GaelVaroquaux We addressed most of the comments and I will push the code with these codes with some tests.

KamalakerDadi commented:
Example which displays Harvard-Oxford, MSDL and Smith 2009 atlas maps:
* Harvard Oxford
```python
import matplotlib.pyplot as plt
from nilearn import datasets, plotting

# Harvard Oxford Atlas
atlas_img_4D = datasets.fetch_harvard_oxford('cort-prob-2mm')
# Visualization of Harvard Oxford Atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0],
                                                title='Harvard-Oxford with contours')
# Visualization of Harvard Oxford Atlas maps as a contour fillings type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], filled=True,
                                                title='Harvard-Oxford with contour fillings')
# Visualization of Harvard Oxford Atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(atlas_img_4D[0], view_type='continuous',
                                                title='Harvard-Oxford with continuous')
plt.show()
```
* MSDL
```python
import matplotlib.pyplot as plt
from nilearn import datasets, plotting
# MSDL Atlas
atlas_data = datasets.fetch_msdl_atlas()
atlas_filename = atlas_data.maps
# Visualization of MSDL Atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(atlas_filename,
                                                title='MSDL with contours')
# Visualization of MSDL Atlas maps as a contour fillings type
display = plotting.img_plotting.plot_prob_atlas(atlas_filename, filled=True,
                                                title='MSDL with contour fillings')
# Visualization of MSDL Atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(atlas_filename, view_type='continuous',
                                                title='MSDL with continuous')
plt.show()
```
* Smith Atlas
```python
import matplotlib.pyplot as plt
from nilearn import datasets, plotting
# Smith 2009 Atlas
smith_data = datasets.fetch_smith_2009()
# Visualization of Smith data atlas maps as a contours type
display = plotting.img_plotting.plot_prob_atlas(smith_data.rsn10,
                                                title='Smith Atlas with contours')
# Visualization of Smith data atlas maps as a contour fillings type
display = plotting.img_plotting.plot_prob_atlas(smith_data.rsn10, filled=True,
                                                title='Smith Atlas with contour fillings')
# Visualization of Smith data atlas maps as a continuous type
display = plotting.img_plotting.plot_prob_atlas(smith_data.rsn10, view_type='continuous',
                                                title='Smith Atlas with continuous')
plt.show()
```
![atlas](https://cloud.githubusercontent.com/assets/11410385/8148392/6e049ec8-129b-11e5-932a-3bbdc0f3c03f.png)

KamalakerDadi commented:
@GaelVaroquaux Pushed some changes. Balance as of now is to add smoke tests.

bthirion commented:
Looks great. Are you providing an example with these displays ?

KamalakerDadi commented:
Yes definitely. I have no specific thing to work on. May be an example should be using canica maps ?

bthirion commented:
Or just the same as you did with atlases ? The advantage is that the example would be faster (no computation involved), which is better if the purpose is to illustrate the visualization capabilities of Nilearn.

GaelVaroquaux commented:
@KamalakerDadi : this is looking good. Now you need to work on the documentation (you need to add an entry to the plotting.rst documentation page) and the examples. It's important to get good wording.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -525,7 +527,13 @@ def add_contours(self, img, **kwargs):
+            # Switches into contour fillings by overriding type as 'contourf'
+            filled_type = 'contour' + 'f'
+        else:
+            filled_type = 'contour'
```
This is shorter and more reable:
```python
filled_type = 'contour'
if filled:
    fikked_type += 'f'

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -525,7 +527,13 @@ def add_contours(self, img, **kwargs):
-        self._map_show(img, type='contour', **kwargs)
+        if filled:
+            # Switches into contour fillings by overriding type as 'contourf'
+            filled_type = 'contour' + 'f'
```
Something like `contour func` would be a better wording for this variable

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,158 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            except ValueError as exc:
+                exc.args += (message, )
+                raise
+            thr = fast_abs_percentile(data, percentile) + 1e-1
```
Don't put magic constants with no reason nor explanation.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,158 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            display.add_contours(map_img, levels=[thr],
+                                 linewidths=linewidths,
+                                 colors=[color])
+            if filled:
```
use the view type `contours_filled` for that case.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,158 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                                     linewidths=linewidths,
+                                     colors=[color[:3].tolist() + [alpha]],
+                                     linestyles='solid', filled=filled)
+        elif view_type == 'continuous':
```
You should raise an error if `view_type` is not an accepted value.

lesteve commented on nilearn/plotting/displays.py:
```diff
@@ -525,7 +527,12 @@ def add_contours(self, img, **kwargs):
                 these contours.
         """
-        self._map_show(img, type='contour', **kwargs)
+        filled_type = 'contour'
```
I personally prefer one-liners for things like that:

```python
filled_type = 'contourf' if filled else 'contour'
```

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -12,7 +12,6 @@
 import functools
 import numbers
 import warnings
-
```
Not such a big deal for this one but please refrain from making changes like this in the future. It just adds unnecessary noise.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                        figure=figure, axes=axes, title=title,
+                        threshold=None, annotate=annotate,
+                        draw_cross=draw_cross, black_bg=black_bg,
+                        cmap=plt.cm.gray, **kwargs)
```
No need to pass explicitly a cmap here just let plot_anat use its default (which is indeed plt.cm.gray)

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                    output_file=None, display_mode='ortho',
+                    figure=None, axes=None, title=None, annotate=True,
+                    draw_cross=True, black_bg='auto', dim=False,
+                    cmap='gist_rainbow', vmin=None, vmax=None,
```
For consistency I would use cmap=plt.cm.gist_rainbow here

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        threshold = "99.7%"
+
+    if hasattr(threshold, '__iter__') and not isinstance(threshold, _basestring):
+        if len(threshold) != n_maps:
```
This is not going to play nicely with the rest of the code if threshold is an iterator. I would just require threshold to be a list for simplicity rather than trying to cater for iterators.

AlexandreAbraham replied:
Result is the same : in any case, iterators are not supported.

AlexandreAbraham replied:
Maybe threshold = list(threshold) before the `if` condtion?

lesteve replied:
if you do require a list, why not use isinstance(threshold, list) then ?

AlexandreAbraham replied:
I think we should go for the simplest solution. We know that the handling of iterators in nilearn is partial and buggy in most functions so we won't handle it in this PR. We just try to support as many types as possible, ie numpy arrays, lists, tuples. I suggest this:

`if hasattr(threshold, '__len__') and not isinstance(threshold, _basestring):`

This matches all previous types but fails for iterators. BTW, if we do that and an iterator is given, it will explode deep in the code with a weird error I think.

lesteve replied:
> BTW, if we do that and an iterator is given, it will explode deep in the code with a weird error I think.

I reckon there should be a catch-all else at the end that raises if threshold is not of one of the accepted types.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            raise TypeError('view_type option should be given '
+                            'either as a "contours" or "contours_filled" '
+                            'or "continuous" ')
+    elif not isinstance(view_type, _basestring):
```
You can replace this line by `else:`.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            The anatomical image to be used as a background. If None is
+            given, nilearn tries to find a T1 template.
+        view_type: string, {'auto', 'contours', 'contours_filled', 'continuous'}, optional
```
Maybe `filled_contours` instead of `contours_filled`?

KamalakerDadi replied:
I just followed the flow of the code. I felt its easy to call both contours and contour fillings when I use the first word naming as ```contours```

AlexandreAbraham replied:
Well, the choice of namings should not be the easiest but the more accurate according to the english language ;). And `filled contours` has much more results than `contours filled` on google.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            display.add_contours(map_img, levels=[thr],
+                                 linewidths=linewidths,
+                                 colors=[color])
+        if view_type == 'contours_filled':
```
You should indent this block.

lesteve replied:
Just curious: .add_contours is going to be called twice if view_type == 'contours_filled' (in which case both `view_type.startswith('contours') and view_type == 'contours_filled'` are True), is that on purpose?

AlexandreAbraham replied:
Yes. `add_countours` plot contours around regions in solid lines. The other mode is used to fill these lines with transparent color.

lesteve replied:
Should the coloring inside the solid lines not happen inside add_contours ?

KamalakerDadi replied:
If I get you correctly. Yes its happening inside provided when we override option ```type``` to ```contourf```

AlexandreAbraham replied:
I think that what @lesteve says is that `add_contours(filled=False)` should draw only the contours and `add_contours(filled=True)` should draw the contours and the filling at the same time, so only one call is required. This is a good point.

KamalakerDadi replied:
I could not addressed this. Any ideas are welcome. Killed most of the time to address this.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                                 colors=[color[:3].tolist() + [alpha]],
+                                 linestyles='solid', filled=True)
+        elif view_type == 'continuous':
+            display.add_overlay(map_img, threshold=float(thr),
```
The `float(...)` was for testing, you can remove it.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        vmax: float
+            Upper bound for plotting, passed to matplotlib.pyplot.imshow
+        alpha: float between 0 and 1
+            Alpha transparency for the brain schematics.
```
This is wrong (probably copied and pasted from plot_glass_brain)

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        # To threshold or choose the level of the contours
+        data = map_img.get_data()
+        if isinstance(thr, _basestring):
+            message = ("If 'threshold' is given as a string it "
```
It seems that it has been copied and pasted from the plot_connectome code. It'd be great to put the functionality in a function and reuse it.

AlexandreAbraham replied:
Agreed. The whole gestion of threshold (value, string, or list of any of them) should be exported in a function.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            For continuous overlays this threshold value serves to select
+            the maps which are greater than a given value or list of given values.
+            If None is given, the maps are thresholded with default value.
+        linewidths: a value, optional
```
"a value" is quite vague

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    display = plot_anat(anat_img, cut_coords=cut_coords,
+                        output_file=output_file, display_mode=display_mode,
+                        figure=figure, axes=axes, title=title,
+                        threshold=None, annotate=annotate,
```
Same thing about leaving threshold unspecified here and letting plot_anat set its default.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            # Append the lower boundary value as 0 for contour fillings
+            display.add_contours(map_img, levels=[thr, 0.],
+                                 linewidths=linewidths,
+                                 colors=[color[:3].tolist() + [alpha]],
```
This seems a bit overcomplicated, is there no way to pass an alpha via the kwargs ?

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                        %type(view_type))
+
+    if isinstance(cmap, _basestring):
+        color_map = plt.cm.get_cmap(cmap)
```
Looks like we require a colormap instance rather than a string for `cmap` for all the other plotting functions. For consistency I would follow the same convention here.

AlexandreAbraham commented:
You should also add your file in `__init__.py`.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            Alpha transparency for the brain schematics.
+    """
+    display = plot_anat(anat_img, cut_coords=cut_coords,
+                        output_file=output_file, display_mode=display_mode,
```
If `output_file` is not None, your function will crash since returned display will be None.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +556,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        if view_type == 'contours_filled':
+            # Append the lower boundary value as 0 for contour fillings
+            display.add_contours(map_img, levels=[thr, 0.],
+                                 linewidths=linewidths,
```
`countourf` does not use `linewidths`. Please remove it to avoid raising unnescessary warning.

KamalakerDadi commented:
ping @AlexandreAbraham @lesteve 

Addressed some of the comments.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -555,6 +590,155 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+
+    if isinstance(threshold, list):
+        if len(threshold) != n_maps:
+            raise TypeError(' The list of values to threshold '
```
Spurious space.

GaelVaroquaux commented:
We need an example.

KamalakerDadi commented:
ping @GaelVaroquaux @AlexandreAbraham 

Added a simple example file. If you need any specific example to be addressed your inputs are welcome.

KamalakerDadi commented:
ping @GaelVaroquaux @AlexandreAbraham @lesteve 

I have also added an entry into plotting.rst. I had to follow the same lines of the previous entries. I hope it goes well.

AlexandreAbraham commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,43 @@
+
+We show here how to plot the atlas maps or statistical maps which are
+4D using a three different display types, for example using
+1. "contours", plots can be seen only in contours
```
I would put "ROIs contours are delineated by colored lines". "plots" have no meaning in any case.

AlexandreAbraham commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,43 @@
+2. "filled_contours", plots can be seen in contours with fillings
+3. "continuous or overlays", plots can be seen in continuous overlays
+
+You can plot each map with each different color which will chose
```
The colors "will be picked randomly in the given colormap". @banilo, could you review the documentation part of this PR at least? You're our expert!

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,43 @@
+
+    Parameters
+    ----------
+    thr: str or a list of strings or number or list of numbers
```
Maybe something like "real values, percentage express in a string or a list of these"

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,43 @@
+    Parameters
+    ----------
+    thr: str or a list of strings or number or list of numbers
+        If threshold is a string itmush finish with a percent sign,
```
itmush -> it must

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,43 @@
 # Core, usage-agnostic functions
 
 
+def check_threshold(thr, data):
```
Just call the parameter threshold rather than thre.

lesteve replied:
One of the point of creating this function was to be able to use it for connectome plotting. I don't think you have done that yet.

lesteve replied:
You should take care of keeping this function generic. For example, the 1e-6 threshold trick should probably not be part of this function because it looks like we don't need it for connectome plotting.

AlexandreAbraham replied:
Second point of this remark has not been addressed.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,43 @@
+            raise ValueError(message)
+        try:
+            percentile = float(thr[:-1])
+            thr = fast_abs_percentile(data, percentile) + 1e-6
```
I see the problem you are addressing but I don't think it's the best way to do it.

Your problem is : when the percentile il too high, all values of the image are kept, including the 0. in the background.
For me, the solution would be to consider the background to be masked by default, because it's background. And this is not specific to the percentile mode. I would remove the `1e-6` and add, just before the return statement:
```python
# Get rid of background values in all cases
thr = max(thr, 1e-6)
```

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,43 @@
+            raise
+        thr = thr
+    elif not isinstance(thr, numbers.Real):
+        raise TypeError('Threshold must be a real value but you '
```
Why did you change the error message? I reckon the original one is more accurate:

```python
'edge_threshold should be either a number '
'or a string finishing with a percent sign'
```

KamalakerDadi commented:
ping @AlexandreAbraham @lesteve 
Added some lines in the ```check_threshold``` to suit for a generic usage and fixed minor documentation changes.

AlexandreAbraham commented on doc/manipulating_visualizing/plotting.rst:
```diff
@@ -40,6 +40,9 @@ different heuristics to find cutting coordinates.
      :target: ../auto_examples/manipulating_visualizing/plot_demo_plotting.html
      :scale: 50
 
+.. |plot_prob_atlas| image:: ../auto_examples/manipulating_visualizing/images/plot_prob_atlas_006.png
```
Your example generates 3 images. So `plot_prob_atlas_006.png` will not exist. Pick an image from `plot_prob_atlas_001.png`, `plot_prob_atlas_002.png` or `plot_prob_atlas_003.png`.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,64 @@
+            exc.args += (message, )
+            raise
+
+        if threshold.endswith('%'):
```
This code has nothing to do here!

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -37,6 +39,50 @@
 # Core, usage-agnostic functions
 
 
+def check_threshold(threshold, data, percentile_calculate, name=None):
```
I think that there is a misunderstanding on this function. You seem to randomly add stuff to address comments without converging toward an acceptable solution. What you must do is see the logic in common between the two functions, see which functional part of the code can be put in common, think about the specifications (or write them down) and then code it.

lesteve replied:
Not sure whether this comment was made before the most recent changes but I don't see too much wrong with this function right now.

I think something which needs to be done is to use it in the plot_connectome code, which hasn't been done yet.

KamalakerDadi commented:
@AlexandreAbraham @GaelVaroquaux @lesteve 
I have added some tests and fixed some pep8 errors. Running nosetests seems to be OK with these new additions in the test files.

AlexandreAbraham commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,47 @@
+We show here how to plot the atlas maps or statistical maps which are
+4D using a three different display types, for example using
+
+1. "contours", ROIs contours are delineated by colored lines.
```
Spurious spaces.

KamalakerDadi replied:
I loooked the preview of the doc. It looks so clumbsy.  Thats the reason why I added some spaces. If you are talking about the line gaps. for example between 1. and 2.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+            The anatomical image to be used as a background. If None is
+            given, nilearn tries to find a T1 template.
+        view_type: {'auto', 'contours', 'filled_contours', 'continuous'}
```
It is optional.

KamalakerDadi replied:
I removed because it exceeded more that 79 according to pep8 warnings.

lesteve replied:
You can go to the next line using \. This is actually the only way to please numpydoc. For an example see:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/manifold/mds.py#L368

AlexandreAbraham replied:
Okay, but you have to escape the newline and properly indent the line below. Thanks for the tip!

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            and also with color fillings inside the contours.
+            If view_type == 'continuous', maps are overlayed as continous
+            colors irrespective of the number maps.
+        threshold: str or list of strings or number or list of numbers, optional
```
Use `str` or `string` but be consistent.

KamalakerDadi replied:
Ok.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -1255,3 +1260,47 @@ def get_slicer(display_mode):
+
+    Parameters
+    ----------
+    threshold: a real value or a percentage in string or list of these.
```
This function does not handle lists of threshold.

KamalakerDadi replied:
We are looping over the list of thresholds in ```plot_prob_atlas``` case where as ```plot_connectome``` I suppose there is no doc for list of thresholds. So, I think we are safe in both cases. Am I wrong ?

AlexandreAbraham replied:
The code is good in both cases. However, this particular function does not handle lists ofr thresholds.

KamalakerDadi replied:
Ok, then at the moment I will remove saying ```list of these```.

KamalakerDadi replied:
Do you think I should add lines of code so that it will also handle list ?

AlexandreAbraham replied:
What I think is that since you've added some tests, your PR is now ready to merge and I don't want to spend weeks arguing about one function. Factorizing this code with the one in `plot_connectome` is not trivial and maybe we should keep that for the sprint.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -1255,3 +1260,47 @@ def get_slicer(display_mode):
+    threshold: a real value or a percentage in string or list of these.
+        For example, if threshold is a percentage expressed in a string
+        it must finish with a percent sign like "99.7%".
+    data: a numpy array of a map or a numpy array of a symmetric matrix.
```
First line of the doc contains only the type of the parameter. Semantic information are described after.

KamalakerDadi replied:
Ok. May be something like this: Will this suits ?
**a numpy array of a map or a adjacency matrix of shape (n, n)**
**data of a map represents the probabilistic atlas of the brain regions whereas adjacency matrix represents the** **edges or connections of the brain regions.**

AlexandreAbraham replied:
There is no map or adjacency matrix specific code here. This is sufficient:
```python
    data: numpy array
        Data on which the threshold will be calculated if the threshold is a percentage.
```

KamalakerDadi replied:
I will stick to this 
```Data on which the threshold will be calculated```
since I already told this in the ```returns```

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -1255,3 +1260,47 @@ def get_slicer(display_mode):
+        For example, if threshold is a percentage expressed in a string
+        it must finish with a percent sign like "99.7%".
+    data: a numpy array of a map or a numpy array of a symmetric matrix.
+    percentile_calculate: define the name of a specific percentile
```
Your code won't work with a string. If you want to handle both `scoreatpercentile` and `fast_abs_percentile`, I would add a `exact=True` parameter rather than passing the function.

KamalakerDadi replied:
I actually checked with ```plot_prob_atlas``` by changing the string of the function. I may also be wrong. Can you please elaborate a bit on this ? ```exact=True```

lesteve replied:
> If you want to handle both scoreatpercentile and fast_abs_percentile, I would add a exact=True parameter rather than passing the function.

Personally I am fine with passing in a function.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                                 colors=[color])
+            if view_type.startswith('filled'):
+                # Append the lower boundary value as 0 for contour fillings
+                display.add_contours(map_img, levels=[thr, 0.],
```
I think I had an old comment somewhere that was saying I would rather have one `add_contours` call and that `add_contours(filled=True)` would do both the filling and the line around it.

KamalakerDadi replied:
Yes I tried with this idea but the only way to happen is to call both ```contour``` and ```contourf``` some where whether it would be in ```draw_2d``` or in the same ```plot_prob_atlas```. It looks complicated for me. Also we need to take care of ```levels``` which needs low level to be appended to ```0.``` to work for contour fillings seems to be bit complicated for me to handle both cases.

AlexandreAbraham replied:
@KamalakerDadi, I find your explanation very shallow!

Basically, adding `filled=True` in `add_contours` is not trivial because some options are common for `contour` and `contourf` and some are not. We would end up with a lot of kwargs in `add_contours` and the code would be much more complicated.

lesteve replied:
The problem in my mind is that the .add_* methods can be called directly by the (advanced) user so they should be convenient to use. Remembering to call add_contours twice plus do the magic with the threshold and the alpha is not a very reasonable expectation.

KamalakerDadi replied:
ping @lesteve @AlexandreAbraham 
Attempts are made regarding this comment. Please let me know if it looks OK.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    maps_img = _utils.check_niimg_4d(maps_img)
+    n_maps = maps_img.shape[3]
+
+    _view_type = set(('auto', 'contours', 'filled_contours', 'continuous'))
```
Just use a clearer name, like allowed_view_types.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    _view_type = set(('auto', 'contours', 'filled_contours', 'continuous'))
+    if isinstance(view_type, _basestring):
+        if view_type not in _view_type:
+            raise TypeError('view_type option should be given '
```
This should be a ValueError rather than a TypeError I reckon.

Also your error message is incomplete, 'auto' is another possibility that you haven't listed.

AlexandreAbraham replied:
:+1: 
Also, you have a set of allowed types. Use this set to generate the error message: it will be shorter and more robust.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,163 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                            'either as a "contours" or "filled_contours" '
+                            'or "continuous" ')
+    else:
+        raise TypeError('view_type option should be '
```
I am not too sure this one is necessary TBH.

The previous check making sure that view_type is one of the allowed view types is enough.

AlexandreAbraham replied:
:+1:
You can use the condition `if not isinstance(view_type, _basestring) or view_type not in _view_type:`.

lesteve commented on nilearn/plotting/img_plotting.py:
```diff
@@ -243,7 +246,7 @@ def get_data(self):
         self.load()
         return self.affine
-    
+
```
Try to avoid fixing PEP8 problems if you don't touch the neighboring lines.

lesteve commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,47 @@
+===========================================================
+
+We show here how to plot the atlas maps or statistical maps which are
+4D using a three different display types, for example using
```
a three -> three

KamalakerDadi replied:
Since we have bug in ```display.add_overlay``` when plotting Harvard Oxford Atlas, can I change the atlas name to MSDL ?

AlexandreAbraham replied:
No, we'll fix the bug.

lesteve commented on nilearn/plotting/tests/test_displays.py:
```diff
@@ -38,3 +41,26 @@ def test_demo_ortho_projector():
     oprojector.close()
+
+
+# Smoke test to check threshold function for plot connectome
```
Remove this comment which doesn't add anything. BTW this is not a smoke test.

lesteve commented on nilearn/plotting/tests/test_displays.py:
```diff
@@ -38,3 +41,26 @@ def test_demo_ortho_projector():
+
+
+# Smoke test to check threshold function for plot connectome
+def test_threshold_plot_connectome():
```
Call it test_check_threshold. This doesn't have anything to do with plot_connectome any more.

lesteve commented on nilearn/plotting/tests/test_displays.py:
```diff
@@ -38,3 +41,26 @@ def test_demo_ortho_projector():
+    wrong_edge_thresholds = ['0.1', '10', '10.2.3%', 'asdf%']
+    for wrong_edge_threshold in wrong_edge_thresholds:
+        assert_raises_regex(ValueError,
+                            'should be a number followed by the percent sign',
```
While you are at it, you should check that `name` is used in the error message. You can use a regexp like:
'{0}.+should be a number followed by the percent sign'.format(name)`

lesteve commented:
I may have missed it but it looks like you are still not using check_threshold in the plot_connectome code.

AlexandreAbraham commented:
We h

AlexandreAbraham commented:
I'll answer that point. In `plot_connectome`, some special calculations are made for the percent case. This is logical as computing a percentile means that we need data to compute the percentile on. I'm still questioning the interest of factorizing the code in a function but I definitely do not want to block the PR just for that. So maybe we can merge it and then find a clean way to factorize the code.

lesteve commented:
> I'll answer that point.

Chatting with @AlexandreAbraham we agreed that check_threshold can be used by the plot_connectome code.

banilo commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,159 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    n_maps = maps_img.shape[3]
+
+    allowed_view_types = list(('auto', 'contours', 'filled_contours', 'continuous'))
+    if not isinstance(view_type, _basestring) or \
```
Aren't additional brackets more pythonic than the backslash here?

GaelVaroquaux replied:
Yes

KamalakerDadi replied:
Fixed

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,159 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            view_type = 'contours'
+        else:
+            view_type = 'continuous'
+    elif view_type == 'contours':
```
`elif` -> `if` otherwise filled will not be properly initialized for `auto`.

KamalakerDadi replied:
Should I change ```elif ``` to ```if``` ?
> otherwise filled will not be properly initialized for auto

We were discussed not to include filled in ```auto```. Sorry I could not get it properly.

AlexandreAbraham replied:
No problem. I may even be the one that told you to put an elif here ;). However, it is weird that you did not spot this problem in the tests (even if they are smoky).

KamalakerDadi replied:
Fixed.

arthurmensch commented on nilearn/plotting/tests/test_displays.py:
```diff
@@ -38,3 +41,26 @@ def test_demo_ortho_projector():
+                        '{0}.+should be either a number or a string'.format(name),
+                        check_threshold,
+                        threshold, adjacency_matrix,
+                        calculate, name)
```
Maybe test that check_threshold returns the right value with valid arguments:
```python
assert( 1. < check_threshold('50%', adjacency_matrix, calculate, name) < 2.)
```

KamalakerDadi replied:
Tests are passed by adding this line. Done

KamalakerDadi commented:
ping @AlexandreAbraham @GaelVaroquaux @lesteve 
I have added an example which visualizes all the 4D atlases. Need some reviews and comments on the updates. If any.

GaelVaroquaux commented:
> I have added an example which visualizes all the 4D atlases.

On my computer, fig 3 of this example (Harvard Oxford as continuous) is
broken. Feel free to drop by to see what I am talking about.

GaelVaroquaux commented:
> > I have added an example which visualizes all the 4D atlases.

Also, I don't really like the fact that we now have two examples, one
named plot_prob_altas.py and the other named plot_probabilistic_atlas.py.
We need to find names that render the difference more obvious.

GaelVaroquaux commented:
You need to import the function in the __init__ of plotting, so that it is directly accessible.

AlexandreAbraham commented on nilearn/plotting/displays.py:
```diff
@@ -516,7 +518,18 @@ def add_contours(self, img, **kwargs):
                 these contours.
         """
+        # Switches into contour filling function 'contourf'
+        #filled_type = 'contourf' if filled else 'contour'
```
Please remove disabled code.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        raise ValueError(message)
+
+    cmap = plt.cm.get_cmap(cmap)
+    # Build a custom colormap for displaying contours
```
There is no custom colormap, it's just a list.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    else:
+        threshold = [threshold] * n_maps
+
+    if view_type == 'contours':
```
`filled = view_type.startsWith('filled')`

banilo commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+Visualizing 4D atlas maps onto the anatomical image
+===========================================================
+
+Basically this example gives an idea on how to visualize the atlas maps which are
```
an idea of

GaelVaroquaux replied:
Remove "Basically". Also "gives an idea". I would say "This example shows how to visualize probabilistic atlases made of 4D images. There are 3 different display types: 1...."

AlexandreAbraham replied:
:+1:

banilo commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+Please see the related documentation for more information to tune between the
+parameters.
+"""
+# Load 4D Atlas maps
```
Comment content redundent with print content

banilo commented on nilearn/plotting/img_plotting.py:
```diff
@@ -25,18 +25,21 @@
 # Core, usage-agnostic functions
 
 
+
```
too many lines (PEP8)

banilo commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        figure : integer or matplotlib figure, optional
+            Matplotlib figure used or its number. If None is given, a
+            new figure is created.
+        axes : matplotlib axes or 4 tuple of float: (xmin, ymin, width, height), optional
```
inconsistency: space before :

banilo commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+               'Smith2009 20 Brainmap', 'Smith2009 70 Brainmap',
+               'ICBM tissues']
+
+for atlas, name in zip(atlas_types, atlas_names):
```
add ``sorted``so that the order will not change different runs of document compilation

banilo commented on nilearn/plotting/displays.py:
```diff
@@ -1258,3 +1254,48 @@ def get_slicer(display_mode):
     return get_create_display_fun(display_mode, PROJECTORS)
+
+
+def check_threshold(threshold, data, percentile_calculate, name):
```
The description of the function is not entirely clear.

banilo commented on nilearn/plotting/displays.py:
```diff
@@ -1258,3 +1254,48 @@ def get_slicer(display_mode):
+
+        threshold = percentile_calculate(data, percentile)
+
+    elif not isinstance(threshold, numbers.Real):
```
One could inverse the if-else here to handle the non-basestring case first, instead of having the body of the function inside an if-clause

banilo commented on nilearn/plotting/img_plotting.py:
```diff
@@ -556,6 +559,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+
+        Parameters
+        ----------
+        maps_img: a nifti like object or the filename of the 4D probabilistic
```
4D prob.. map [no s]..because you already said it is 4d

GaelVaroquaux commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+
+3. "continuous", maps are shown as just color overlays.
+
+This function can display each map with each different color which are picked
```
"This function"? Which function? I would say:

"The :func:`nilearn.plotting.plot...` function displays..."

GaelVaroquaux commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+               'Smith2009 10 RSNs', 'Smith2009 20 RSNs',
+               'Smith2009 70 RSNs', 'Smith2009 10 Brainmap',
+               'Smith2009 20 Brainmap', 'Smith2009 70 Brainmap',
+               'ICBM tissues']
```
Here I would like a dictionary to be used.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -76,6 +79,51 @@ def _get_plot_stat_map_params(stat_map_img, vmax, symmetric_cbar, kwargs, force_
+    threshold: a real value or a percentage in string or list of these.
+        For example, if threshold is a percentage expressed in a string
+        it must finish with a percent sign like "99.7%".
+    data: a numpy array of a map or a numpy array of a symmetric matrix.
```
You are missing a description of what is data here (you're just describing it's type).

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -595,6 +643,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+                    draw_cross=True, black_bg='auto', dim=False,
+                    cmap=plt.cm.gist_rainbow, vmin=None, vmax=None,
+                    alpha=0.5, **kwargs):
+
```
This empty line shouldn't be there.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -595,6 +643,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+        Parameters
+        ----------
+        maps_img: a nifti like object or the filename of the 4D probabilistic
+            atlas maps or statistical maps
```
Break down this entry in the type, which goes on the first line, and what it means (4D probabilistic...) that goes on a line below.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -595,6 +643,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+    maps_img = _utils.check_niimg_4d(maps_img)
+    n_maps = maps_img.shape[3]
+
+    allowed_view_types = list(('auto', 'contours', 'filled_contours', 'continuous'))
```
You haven't addressed my comment that I did earlier:

- Rename 'allowed_view_types' to 'valid_view_types'

- Don't use "list(('a', 'b', 'c'))" but "['a', 'b', 'c']"

AlexandreAbraham replied:
Sorry, we had to deal with a git spaghetti history and lost comments in the process.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -595,6 +643,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            correction_factor = .5
+        threshold = "%f%%" % (100 * (1 - .2 * correction_factor / n_maps))
+
+    if isinstance(threshold, list):
```
You don't want to be testing for a list, because that's too restrictive. You need to use something like:

<pre>
import collections

if isinstance(e, collections.Iterable):
</pre>

AlexandreAbraham replied:
You need at least to check that `threshold` is not a `_basestring`.

GaelVaroquaux replied:
> You need at least to check that threshold is not a _basestring.

You have a point. But let's at least accept tuples and arrays, or there
will be a surprising behavior.

lesteve replied:
Also supporting iterables in general, without an extremely good reason, is a pain in the ass as we already experienced elsewhere in nilearn. For example, this code will break with an iterable because you consume the iterable at least twice, `len(threshold) and `enumerate(zip(..., threshold))`.

GaelVaroquaux replied:
Yes, it's fine. Let's just support list, tuple and arrays, which are the
comon cases.

GaelVaroquaux commented:
A bunch of comments to address above.

In addition, the function should be added to the reference documentation: doc/modules/reference.rst

AlexandreAbraham commented on doc/manipulating_visualizing/plotting.rst:
```diff
@@ -217,8 +225,8 @@ plot, and has methods to add overlays, contours or edge maps::
-                  |hack|
-                  **Example:** :ref:`example_manipulating_visualizing_plot_probabilistic_atlas.py`
+                   |hack|
+                   **Example:** :ref:`example_manipulating_visualizing_plot_overlay.py`
```
Extra spaces.

AlexandreAbraham commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+randomly from the colormap which is already defined.
+
+Please see the related documentation for more information to tune between the
+parameters.
```
I think that there is no `between` after `tune`.

AlexandreAbraham commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,60 @@
+
+for atlas, name in zip(atlas_types, atlas_names):
+        plotting.plot_prob_atlas(atlas,
+                                              title='%s' % name)
```
Unnecessary newline.

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -76,6 +79,51 @@ def _get_plot_stat_map_params(stat_map_img, vmax, symmetric_cbar, kwargs, force_
     return cbar_vmin, cbar_vmax, vmin, vmax
 
+
+def check_threshold(threshold, data, percentile_calculate, name=None):
```
This function is in `displays.py` and in `img_plotting.py`. You only need one version ;).

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -76,6 +79,51 @@ def _get_plot_stat_map_params(stat_map_img, vmax, symmetric_cbar, kwargs, force_
+    ----------
+    threshold: a real value or a percentage in string or list of these.
+        For example, if threshold is a percentage expressed in a string
+        it must finish with a percent sign like "99.7%".
```
Missing newline

AlexandreAbraham commented on nilearn/plotting/img_plotting.py:
```diff
@@ -675,12 +885,12 @@ def plot_stat_map(stat_map_img, bg_img=MNI152TEMPLATE, cut_coords=None,
     bg_img, black_bg, bg_vmin, bg_vmax = _load_anat(bg_img, dim=dim,
                                                     black_bg=black_bg)
-    
+
```
For next PR: try not to fix PEP8 errors in random code. Only fix them if you work in the concerned code.

GaelVaroquaux commented on nilearn/plotting/img_plotting.py:
```diff
@@ -595,6 +600,168 @@ def plot_roi(roi_img, bg_img=MNI152TEMPLATE, cut_coords=None,
+            colors irrespective of the number maps.
+        threshold: None, a str or a number, list of either str or number, optional
+            If threshold is a string it must finish with a percent sign,
+            e.g. "25.3%", or if it is a number it should be real numbers.
```
To be more precise, I would write.

e.g. "25.3%", and it is a percentile. Or if it is a number, if should be a real number, in which case it is the value to threshold at.

GaelVaroquaux commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,57 @@
+"""
+Visualizing 4D probabilistic atlas maps onto the anatomical image
```
I would make this title shorter:

"Visualizing 4D probabilistic atlas maps"

GaelVaroquaux commented on examples/manipulating_visualizing/plot_prob_atlas.py:
```diff
@@ -0,0 +1,57 @@
+               'Smith2009 70 Brainmap': smith.bm70,
+               'ICBM tissues': (icbm['wm'], icbm['gm'], icbm['csf'])}
+
+for name, atlas in atlas_types.items():
```
You need a 'sorted' here, because 'items()' will not always return the values in the same order. So the code would be:

"for name, atlas in sorted(atlas_types.items()):"

GaelVaroquaux commented:
This is gr8.

I want to merge.

@AlexandreAbraham : can I merge?

lesteve commented:
I think @KamalakerDadi is still tweaking a few things (sphinx minor issues mostly).

GaelVaroquaux commented:
> I think @KamalakerDadi is still tweaking a few things (sphinx minor issues
> mostly).

Correct. But is this a :+1: once he is done?

AlexandreAbraham commented:
You have my blessing :pray:

GaelVaroquaux commented:
Merging! (wild!)
